File: part 3/practice/Interpreter.py
from Token import Token


class Interpreter(object):
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_token = None
        self.current_char = self.text[self.pos]

    def error(self):
        raise Exception('Error parsing input')

    def integer(self):
        number = ''  # self.current_char

        while self.pos < len(self.text) and self.current_char.isdigit():
            number += self.current_char if self.current_char and self.current_char.isdigit() else ''
            self.next()

        return int(number)

    def next(self):
        self.pos += 1
        self.current_char = self.text[self.pos] if self.pos < len(self.text) else None

    def skip(self):
        while self.current_char and self.current_char.isspace():
            self.next()

    def get_next_token(self):
        """Lexical analyzer (also known as scanner or tokenizer)

        This method is responsible for breaking a sentence
        apart into tokens. One token at a time.
        """

        if self.pos > len(self.text) - 1:
            return Token(Token.EOF, None)

        if self.current_char.isdigit():
            return Token(Token.INTEGER, self.integer())

        if self.current_char == '*':
            self.next()
            return Token(Token.MULTIPLY, '*')

        if self.current_char == '/':
            self.next()
            return Token(Token.DIVIDE, '/')

        if self.current_char.isspace():
            self.skip()
            return self.get_next_token()

        self.error()

    def eval(self, token_types):
        """compare the current token type with the passed token family"""

        if self.current_token.type in token_types:
            self.current_token = self.get_next_token()

        else:
            self.error()

    def op(self, left, op, right):
        """Arithmetic operator dispatcher."""
        if op == '*':
            return left * right
        elif op == '/':
            return left / right

    def opt(self):
        """<opt> ::= <op> <term>"""

        op = self.current_token
        self.eval([Token.MULTIPLY, Token.DIVIDE])

        right = self.current_token
        self.eval([Token.INTEGER])

        return op, right

    def expr(self):
        """<expression> ::= <term> { <op> <term> }*"""

        self.current_token = self.get_next_token()

        result = self.current_token.value
        self.eval([Token.INTEGER])

        while self.current_token.type != Token.EOF:
            op, right = self.opt()
            result = self.op(result, op.value, right.value)

        return result

File: part 3/practice/Interpreter.cs
using System;

public class Interpreter
{
	int Position { get; set; }
	string Text { get; init; }
	int Len { get { return this.Text.Length; } }

	List<Token> Tokens { get; set; }

	public Interpreter(string text)
	{
        Text = text;
		Tokens = new List<Token>();
	}

	private void Skip()
	{
		while (Char.IsWhiteSpace(Text[Position+1])) { Position++; }
	}

	private void Next()
	{
		if (Position < Len)
		{
			Position++;
		}
	}

	private Token TokenizeInt()
	{
		string number = Text[Position].ToString();

		while (Position < Len - 1 && Char.IsDigit(Text[Position+1]))
		{
			number += Text[++Position];
		}

		//Position++;

		return new Token(Token.INT, number);
	}

	private void Error()
	{
		throw new Exception($"Invalid element exception. {Text[Position]} was not expected.");
	}

	public void Tokenize()
	{
		while (Position < Len)
		{
			switch (Text[Position])
			{
				case ' ':
					Skip();
					break;

				case '+':
                    Tokens.Add(new Token(Token.PLUS, "+"));
					break;

				case '-':
                    Tokens.Add(new Token(Token.MINUS, "-"));
					break;

				default:
					if (Char.IsDigit(Text[Position]))
					{
						Tokens.Add(TokenizeInt());
					} else
					{
						Error();
					}
					break;

			}

            Next();
        }
    }

	private int Op(int left, string op, int right)
	{
		switch (Char.Parse(op))
		{
			case '+':
				return left + right;

			case '-':
				return left - right;

			default:
				Error();
				return -1;
		}
	}

	public int Eval()
	{
		Tokenize();

		int result = int.Parse(Tokens[0].Value);

		for (int i = 1; i < Tokens.Count(); )
		{
			result = Op(result, Tokens[i++].Value, int.Parse(Tokens[i++].Value));
		}

		return result;
	}
}
